sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,652 @@ These follow the AAELF64_ base definition from Arm.
+    autia1716
+    br   x17
+
+The PAC PLT requires the dynamic linker to sign the entry for
+``.got.plt[N]`` with the address of ``.got.plt[N]`` as the modifier.
+
+PAC + BTI
+~~~~~~~~~
+
+This is a combination of the PAC and the BTI PLT entries. The static
+linker must set both ``DT_AARCH64_BTI_PLT`` and ``DT_AARCH64_PAC_PLT``
+when BTI compliant PLT entries requiring signed ``.got.plt``
+entries are used.
+
+PLT header ``PLT[0]``
+
+.. code-block:: asm
+
+    bti  c
+    stp  x16, x30, [sp,#-16]!
+    adrp   x16, :page: &.got.plt[2]
+    ldr    x17, [x16, :lo12: &.got.plt[2]]
+    add    x16, x16, :lo12: &got.plt[2]
+    br   x17
+
+Nth PLT entry ``PLT[N]``
+
+.. code-block:: asm
+
+    bti  c
+    adrp x16, :page: &.got.plt[N + 3]
+    ldr  x17, [x16, :lo12: &.got.plt[N + 3]]
+    add  x16, x16, :lo12: &.got.plt[N + 3]
+    autia1716
+    br   x17
+
+GNU Indirect Functions
+----------------------
+
+A GNU Indirect Function (IFUNC) is a feature that permits a single
+implementation of a function to be chosen from multiple candidates,
+with the choice taken by an IFUNC resolver function.
+
+GNU Indirect Functions require static and dynamic linker support. They
+are known to be supported on GNU/Linux, Android, and many of the BSD
+operating systems.
+
+GNU Indirect Functions are called via a PLT entry that loads the
+function address. The function address is chosen by an IFUNC resolver
+function.
+
+The source code interface to an IFUNC resolver is platform
+dependent. The GNU/Linux interface via the GNU C Library is documented
+below.
+
+GNU C Library IFUNC interface
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The GNU C Library (glibc) ``sys/ifunc.h`` header provides the
+necessary type definitions for defining IFUNC resolvers. The resolver
+is passed at least one, and at most 2 parameters dependent on the
+glibc IFUNC ABI version. A flag is set in the first parameter if a
+second parameter is available. The parameters provide access to the
+HWCAP_ flags.
+
+.. code-block:: c
+
+  /* A second argument is passed to the ifunc resolver.  */
+  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+
+  /* Function prototype for an IFUNC resolver.
+  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+
+  /* Second argument to an ifunc resolver.  */
+  struct __ifunc_arg_t
+  {
+    unsigned long _size; /* Size of the struct, so it can grow.  */
+    unsigned long _hwcap;
+    unsigned long _hwcap2;
+  };
+
+  typedef struct __ifunc_arg_t __ifunc_arg_t;
+
+IFUNC resolver functions must have a type of ``STT_GNU_IFUNC``. With
+the GCC and Clang compilers an attribute can be used to achieve this.
+
+.. code-block:: c
+
+  /* Implementation alternative 1 */
+  static int implementation1(void) { ... }
+
+  /* Implementation alternative 2 */
+  static int implementation2(void) { ... }
+
+  /* Resolver function */
+  static void* resolver(uint64_t, const __ifunc_arg_t *) { ... }
+
+  /* make symbol ifunc type STT_GNU_IFUNC using resolver as the IFUNC resolver. */
```

Comment:
s/make/Make/

Reply:
It is meant to represent that the symbol index can be 0. I'll try and clarify. In practice I've not seen any with a non-0 index but the case is well defined in the ABI. Quotiing from  Delta(S) in https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst#5733relocation-operations .

> Delta(S) if S is a normal symbol, resolves to the difference between the static link address of S and the execution address of S. If S is the null symbol (ELF symbol index 0), resolves to the difference between the static link address of P and the execution address of P.